test/node-unit/worker.spec.js

```diff
@@ -0,0 +1,199 @@
+'use strict';
+
+const serializeJavascript = require('serialize-javascript');
+const {SerializableWorkerResult} = require('../../lib/serializer');
+const rewiremock = require('rewiremock/node');
+const {createSandbox} = require('sinon');
+
+const WORKER_PATH = require.resolve('../../lib/worker.js');
+
+describe('worker', function() {
+  let worker;
+  let sandbox;
+  let stubs;
+
+  beforeEach(function() {
+    sandbox = createSandbox();
+    stubs = {
+      workerpool: {
+        isMainThread: false,
+        worker: sandbox.stub()
+      }
+    };
+    sandbox.spy(process, 'removeAllListeners');
+  });
+
+  describe('when run as main process', function() {
+    it('should throw', function() {
+      expect(() => {
+        rewiremock.proxy(WORKER_PATH, {
```

Comment:
I don't like these kinds of "powermock" magic. It makes it pretty unclear for contributors to know what's happening. Could we maybe refactor `worker.js` to know if were running in a child process a different way? I would like to take a shot at it, if you agree.

Reply:
It's true that it's not immediately obvious what's happening here.  However, module-level mocking using `rewiremock` is much safer than simply stubbing out functions or properties, and offers better isolation.  If I had the time, I'd use it in _all_ of our unit tests--some subtle and not-so-subtle bugs in the tests can be introduced if stubbing with `sinon` alone.

In this case, though, I believe unfamiliarity is misconstrued for complexity.  

Perhaps wrapping rewiremock in a more obvious API would be a good idea though.  And I'm happy to comment what exactly is happening here, because, yeah, it's a little weird.  In a nutshell:

1. When you execute `worker.js` via `node`--_not_ loading it via `workerpool`--it will immediately throw an exception to let the developer know they screwed up.  This is the unusual thing.
1. `worker.js` checks to see if it's been loaded via `require('workerpool').isMainThread`.
1. It is safer to stub out `workerpool` in its entirety than to monkeypatch `isMainThread` in `workerpool`.  
1. So, we use `rewiremock` to essentially `require('worker.js')`, except instead of letting it use the _real_ `workerpool` module, we give it a phony object that has an `isMainThread` prop set to `true`.  This does not affect any other use of `workerpool`, nor any other use of `worker.js`.

Reply:
In this case,  I would personally opt to not check the `isMainThread` property. Instead I would make `worker.js` safe to load from the main process, but make sure it doesn't have any side effects. This is what I did in Stryker: https://github.com/stryker-mutator/stryker/blob/master/packages/core/src/child-proxy/ChildProcessProxyWorker.ts#L125. If devs screw up, you won't get an error (don't feel strongly about this myself).

It's ok if we agree to disagree here. 

>  If I had the time, I'd use it in all of our unit tests--some subtle and not-so-subtle bugs in the tests can be introduced if stubbing with `sinon` alone.

I do agree with the statement about `sinon` alone. I'm a big fan of dependency injection. Instead of monkey-patching, you would be able to just provide a mock in the constructor/function. However, that would mean a bigger rewrite and is not realistic 😉